Initial implementation of ForwardCRD plugin

christianang opened the pull request:
### 1. Why is this pull request needed and what does it do?

This PR is the initial implementation of the KubernetesCRD plugin that sets out to implement #4382. Namely, it allows CoreDNS to watch a DNSZone CRD to configure stub-domains for Kubernetes clusters.

Edit: This PR has been updated from just the readme to the initial implementation of the plugin.

### 2. Which issues (if any) are related?

#4382

### 3. Which documentation changes (if any) need to be made?

### 4. Does this introduce a backward incompatible change or deprecation?

rosenhouse commented on plugin/kubernetescrd/README.md:
```diff
@@ -0,0 +1,161 @@
+# kubernetescrd
```
Small note from reviewing this README:  maybe a few words or a section about security?

A malicious user could create these resources in order to steal traffic (as they could if they had write access to the `coredns` configmap in `kube-system`

May be worth recommending that users limit the namespace to `kube-system`, perhaps show that example first with some language recommending that?  Also worth noting that when touching on the RBAC topic, since RBAC on writes to the resource could be another way of locking down the security.

christianang replied:
Good point. I've added a small section to readme wrt to this.

chrisohaver commented on core/dnsserver/zdirectives.go:
```diff
@@ -46,6 +46,7 @@ var Directives = []string{
 	"clouddns",
 	"k8s_external",
 	"kubernetes",
+	"kubernetescrd",
```
The position of this plugin in the chain is driven by the DNS function performed, which for now is just forwarding.  If the plugin is expanded to expose other functions via CRDs, then picking a single position in the plugin chain that is appropriate for all functions may prove difficult/impossible. IOW, this plugin may need to remain specific to stub domain forwarding.

We might want to order this immediately before _forward_. In its current position, it executes before _file_, whereas _forward_ normally executes after _file_.  I think we should try to make the behavior be as close as possible to using the _forward_ plugin.

However, the typical way of configuring a stub domain is to define a serverblock for the zone.  A separate serverblock creates a plugin chain for that zone, allowing for different configurations per stubdomain  (e.g. cache settings).  Routing all the zones from a single point isn't quite as flexible. There is currently no way for plugins to "create" new server blocks.  I think it would be possible to modify core/dnsserver to enable that, but it could get complicated.

chrisohaver replied:
Another difference between routing queries through separate serverblocks vs single point will surface in metrics, since server scope metrics have a zone label of the enclosing serverblock.  When routing queries through one point, the resulting metrics would all be labeled with the same zone (the zone of the serverblock enclosing the plugin).

chrisohaver replied:
> If the plugin is expanded to expose other functions via CRDs, then picking a single position in the plugin chain that is appropriate for all functions may prove difficult/impossible.

Regarding the above: This may be solved by instead of having the crd-monitoring plugin serving DNS requests, the other plugins did.  For example, if forward could handle multiple zones, learn of added/deleted zones from the crd-monitoring plugin while running, and start/stop serving those zones.  In this way, dns requests would be served by the forward plugin itself.  Other plugins could be adapted similarly, e.g. hosts, for crd controlled A/AAAA records.  The hosts and forward plugins require different positions in the plugin chain (hosts near the top, and forward near the bottom).

Note: this doesn't address the aforementioned "serverblock" differences.

christianang replied:
> The position of this plugin in the chain is driven by the DNS function performed, which for now is just forwarding. If the plugin is expanded to expose other functions via CRDs, then picking a single position in the plugin chain that is appropriate for all functions may prove difficult/impossible. IOW, this plugin may need to remain specific to stub domain forwarding.

Positioning in the chain is an interesting point I didn't realize. I was optimistically aiming for this plugin to encapsulate other functions via CRDs, but it does seem difficult now. Perhaps it is best to rename this plugin if you think it is unlikely we can encapsulate other functionality into this plugin (maybe `forwardingcrd` or something)? It does feel like it may get unwieldy if there becomes a proliferation of plugins to implement more functionality via CRDs.

> We might want to order this immediately before forward. In its current position, it executes before file, whereas forward normally executes after file. I think we should try to make the behavior be as close as possible to using the forward plugin.

Makes sense, and I can make this change.

> However, the typical way of configuring a stub domain is to define a serverblock for the zone. A separate serverblock creates a plugin chain for that zone, allowing for different configurations per stubdomain (e.g. cache settings). Routing all the zones from a single point isn't quite as flexible. There is currently no way for plugins to "create" new server blocks. I think it would be possible to modify core/dnsserver to enable that, but it could get complicated.

Yeah, the current implementation is not very flexible in this regard. I like the idea of this plugin "creating" new server blocks, but it does seem very complex and I did not want to dig too deep into modifying core/dnsserver code on the initial implementation. I could see the plugin being refactored to use such functionality and perhaps the CRD at that point could allow a user to specify what plugins they want to enable and those could be translated into server blocks. Perhaps creating new server blocks would also helps with the plugin chain ordering problem and the metrics problem?

> Regarding the above: This may be solved by instead of having the crd-monitoring plugin serving DNS requests, the other plugins did. For example, if forward could handle multiple zones, learn of added/deleted zones from the crd-monitoring plugin while running, and start/stop serving those zones. In this way, dns requests would be served by the forward plugin itself. Other plugins could be adapted similarly, e.g. hosts, for crd controlled A/AAAA records. The hosts and forward plugins require different positions in the plugin chain (hosts near the top, and forward near the bottom).

That makes sense, but one thing I'm unclear on, do instances of plugins just exist that I can modify (assuming we gave that plugin the functionality to be externally modified), or would the crd-monitoring plugin still have to instantiate instances of the plugin (and would that mean it isn't part of the proper plugin chain)?

chrisohaver replied:
> Perhaps creating new server blocks would also helps with the plugin chain ordering problem and the metrics problem?

Yes - it would essentially mimic the current way stub zones are configured.  So, metrics would get labeled in the correct zone.  Plugin chain ordering would not matter, since each zone gets its own chain.

I've been looking into how this could be implemented, and what it looks like.  After a few hours fiddling with it, it can confirm, it _is_ complicated - and requires some (perhaps unacceptable) hacks.  ATM, I'm thinking that my tinkering wont produce something feasible, and we'll just document the differences in behavior between CRD stub zones and manually defined ones.

> That makes sense, but one thing I'm unclear on, do instances of plugins just exist that I can modify (assuming we gave that plugin the functionality to be externally modified), or would the crd-monitoring plugin still have to instantiate instances of the plugin (and would that mean it isn't part of the proper plugin chain)?

I was thinking we assume they exist.  Dynamically inserting another plugin into the chain while maintaining the correct execution order would almost impossible I think.  Assuming the plugin exists means we'd probably want the dns-serving plugins to be declared in a blank slate state (to later be configured via the crd interface). But in the CRD stub zone forward case, an instance of forward typically pre-exists (the default upstream).

The crd plugin would not sit in the plugin chain. It would not have a `ServeDNS()` function defined (wont implement a plugin Handler), and it would not be added to the Config in setup().  There are examples of this kind of "configuration" type plugin, e.g. *sign*

Another approach entirely could be to create "crd" versions of each plugin we want to CRD-ify... i.e. *forward-crd* , *hosts-crd*, ... and package some of the shared crd monitoring components to reduce code duplication. These crd plugins could be placed next to (just before, just after?) their non-crd counterparts. *forward-crd* essentially akin to the implementation in this PR.  This minimizes ordering angst, but i think there may still be a little WRT the before/after question (i.e. which one "wins" in the event of zone overlap).

christianang replied:
Any preference on which implementation is preferable? Not sure if I lean towards any particular implementation since they both seem to have their own tradeoffs. Configuring a separate plugin means you have to add both plugins to the chain, which could lead to misconfiguration, but solves the ordering issue. CRD-ifying plugins solves adding both plugins to the chain, but there is now an ordering issue with their non-crd counterparts.

chrisohaver replied:
> Any preference on which implementation is preferable?

I like the "crd plugin configures other plugins" solution because it mimics the behavior of having had configured those plugins via the Corefile, and it results in fewer plugins down the road.
However, I like the "forward-crd plugin serves dns itself via an instance of forward" because it would enable other plugins to leverage the forwarding capability of the forward plugin.

I'm leaning toward "forward-crd plugin serves dns itself via an instance of forward" ... since i'm on the fence, and you've more or less already done it this way.

christianang replied:
Sounds good. I've done a quick rename from `kubernetescrd` to `forwardcrd`. I think that is all I might need to do at this point, but let me know if I've missed anything.

chrisohaver commented on plugin/forward/forward.go:
```diff
@@ -53,12 +56,112 @@ type Forward struct {
 
+// NewWithConfig returns a new Forward configured by the provided
+// ForwardConfig.
+func NewWithConfig(config ForwardConfig) (*Forward, error) {
```
I think I like this direction.  I presume forward's existing Corefile parsing functions will also be refactored to create a `ForwardConfig`, and then use `NewWithConfig` to create the `Forward` struct.  This also opens the way for other plugins to reuse the "forwarding" function of the forward plugin (e.g. adding a forwarding action to coredns/policy).

@miekg, WDYT?

v-byte-cpu replied:
@chrisohaver I love this idea and implemented a similar solution in https://github.com/coredns/coredns/pull/4583. Anyway config parsing logic should be separated from plugin logic according to the single-responsibility principle. It will not only decouple components but also allow to thoroughly cover the parsing functionality with tests.

chrisohaver commented on plugin/kubernetescrd/controller.go:
```diff
@@ -0,0 +1,235 @@
+		if err != nil {
+			return err
+		}
+		err = plugin.OnStartup()
```
This doesnt capture the dnstap integration (from forward setup).  Not sure if it's achievable from this point.

christianang replied:
Good catch, but what makes you say you don't think it is achievable? I think I can save off the `dnstap` plugin within setup and set the `tapPlugin` on `forward.NewWithConfig`, but let me know if there is something I'm missing.

chrisohaver replied:
Oh it could be - I didn't scratch below the surface ...  just saying I wasn't sure if it was or not.

christianang replied:
Gotcha, I implemented dnstap and tested it locally.

chrisohaver reviewed (commented):
Thanks @christianang! Lots to digest, but from what I understand, it looks good so far.

chrisohaver commented on plugin/forwardcrd/README.md:
```diff
@@ -0,0 +1,173 @@
+## Name
+
+*forwardcrd* - enables proxying DNS messages to upstream resolvers by reading
+CRDs from a Kubernetes cluster
```
```suggestion
the `DNSZone` CRD from a Kubernetes cluster
```

chrisohaver commented on plugin/forwardcrd/README.md:
```diff
@@ -0,0 +1,173 @@
+
+This plugin reports readiness to the ready plugin. This will happen after it has
+synced to the Kubernetes API.
+
```
Should add a section (here or elsewhere) describing how this works with and differs from the Corefile serverblock defined stub domains, and a forward plugin in the same block that covers the following points...
* when DNSZone CRDs and serverblock defined stub domains are used, domains defined in the Corefile take precedence (in the event of zone overlap). e.g. if the domain `example.com` is defined in the corefile as a stub domain, and a DNSZone CRD record defined for `sub.example.com`, then `sub.example.com` would get forwarded to the upstream defined in the corefile, not the DNSZone.
* metrics difference: DNSZone metrics all labeled in single zone (the zone of the enclosing serverblock).
* When using forwardcrd and forward in the same serverblock, DNSZones take precedence over forward plugin defined in the same serverblock. e.g. if a DNSZone is defined for `.`, then no queries would be forwarded to the upstream defined the forward plugin of the same serverblock.

precedence is serverblock -> forwardcrd -> forward

christianang replied:
Makes sense. Added to the readme.

chrisohaver reviewed (commented):
> I added a function to the forward plugin NewWithConfig that allows the kubernetescrd plugin to create instances of forward on-demand. This duplicated some code, which could be "DRY"-ed up, but in favor of not making too far reaching changes to the code base I left it duplicated and am happy with making follow up PRs to de-duplicate code if the approach I used makes sense.

I think it makes sense to work that out now and include it in this PR.

chrisohaver commented on plugin/forwardcrd/README.md:
```diff
@@ -0,0 +1,173 @@
+  ignored if connecting in-cluster (i.e., the endpoint is not specified).
+* `namespace` **NAMESPACE** only reads `DNSZone` resources from the namespace
+  listed. If this option is omitted then it will read from all namespaces.
+  **Note**: It is recommended to limit the namespace (e.g to `kube-system`)
```
Ideally we should default to the recommended configuration.  e.g. Default to only reading CRDs from `kube-system`, then perhaps allowing just `namespace` without any namespaces listed mean "all namespaces".

christianang replied:
Makes sense.

flawedmatrix committed:
Refactor forward setup to use NewWithConfig

Co-authored-by: Christian Ang <angc@vmware.com>

Signed-off-by: Edwin Xie <exie@vmware.com>

chrisohaver commented on plugin/forwardcrd/README.md:
```diff
@@ -0,0 +1,206 @@
+
+When `DNSZone` CRDs and Server Blocks define stub domains that are used,
+domains defined in the Corefile take precedence (in the event of zone overlap).
+in the Corefile take precedence (in the event of zone overlap). e.g. if the
```
```suggestion
e.g. if the
```

chrisohaver commented on plugin/forwardcrd/README.md:
```diff
@@ -0,0 +1,206 @@
+
+## Ordering
+
+Coredns has following precedence:
```
```suggestion
Forward behavior can be defined in three ways, via a Server Block, via the *forwardcrd* plugin, and via the *forward* plugin.  If more than one of these methods is employed and a query falls within the zone of more than one, CoreDNS selects which one to use based on the following precedence:
```

chrisohaver commented on plugin/forwardcrd/README.md:
```diff
@@ -0,0 +1,206 @@
+## Ordering
+
+Coredns has following precedence:
+Corefile Server Block -> `forwardcrd` plugin -> `forward` plugin.
```
```suggestion
Corefile Server Block -> *forwardcrd* plugin -> *forward* plugin.
```

chrisohaver replied:
To be consistent with other plugin documentation, *emphasize* plugin names.

chrisohaver reviewed (commented):
LGTM!  Suggested some fixes and tweaks to docs.

chrisohaver commented on plugin/forwardcrd/README.md:
```diff
@@ -0,0 +1,207 @@
+When using `forwardcrd` and `forward` in the same Server Block, `Forward` CRDs
+take precedence over the `forward` plugin defined in the same Server Block.
+e.g. if a `Forward` resource is defined for `.`, then no queries would be
+forwarded to the upstream defined the `forward` plugin of the same Server Block.
```
```suggestion
When using *forwardcrd* and *forward* plugins in the same Server Block, `Forward` CRDs
take precedence over the *forward* plugin defined in the same Server Block.
e.g. if a `Forward` CRD is defined for `.`, then no queries would be
forwarded to the upstream defined in the *forward* plugin of the same Server Block.
```

chrisohaver commented on plugin/forwardcrd/README.md:
````diff
@@ -0,0 +1,207 @@
+kubectl apply -f ./manifests/crds/coredns.io_forwards.yaml
+```
+
+Assuming the **forwardcrd** plugin has been configured to allow `Forward`
````
```suggestion
Assuming the *forwardcrd* plugin has been configured to allow `Forward`
```

chrisohaver commented on plugin/forwardcrd/README.md:
```diff
@@ -0,0 +1,207 @@
+`Forward` CRD metrics are all labeled in a single zone (the zone of the enclosing
+Server Block).
+
+## Examples
```
A real-world example of how forwardcrd would be used in an actual CoreDNS Corefile as deployed in a kubernetes cluster would be helpful.

chrisohaver commented on plugin.cfg:
```diff
@@ -61,6 +61,7 @@ secondary:secondary
 etcd:etcd
 loop:loop
 forward:forward
+forwardcrd:forwardcrd
```
_forwardcrd_ should come before _forward_. Otherwise, a default upstream forwarder, such as the one in the default K8s file `forward . /etc/resolv.conf`, would prevent _forwardcrd_ in the same server block from serving any queries.

christianang committed:
Add forwardcrd plugin README.md

Co-authored-by: Aidan Obley <aobley@vmware.com>

Signed-off-by: Christian Ang <angc@vmware.com>

christianang committed:
Create forwardcrd plugin

- Place forwardcrd before forward plugin in plugin list. This will avoid
forward from preventing the forwardcrd plugin from handling any queries
in the case of having a default upstream forwarder in a server block (as
is the case in the default kubernetes Corefile).

Co-authored-by: Aidan Obley <aobley@vmware.com>

Signed-off-by: Christian Ang <angc@vmware.com>

christianang committed:
Add Forward CRD

Signed-off-by: Christian Ang <angc@vmware.com>

christianang committed:
Add NewWithConfig to forward plugin

- allows external packages to instanciate forward plugins

Co-authored-by: Aidan Obley <aobley@vmware.com>

Signed-off-by: Christian Ang <angc@vmware.com>

christianang committed:
ForwardCRD plugin handles requests for Forward CRs

- add a Kubernetes controller that can read Forward CRs
- instances of the forward plugin are created based on Forward CRs from
the Kubernetes controller
- DNS requests are handled by calling matching Forward plugin instances
based on zone name
- Defaults to the kube-system namespace to align with Corefile RBAC

Signed-off-by: Christian Ang <angc@vmware.com>

Use klog v2 in forwardcrd plugin

christianang committed:
Use ParseInt instead of Atoi

- to ensure that the bitsize is 32 for later casting to uint32

Signed-off-by: Christian Ang <angc@vmware.com>

christianang committed:
Add @christianang to CODEOWNERS for forwardcrd

Signed-off-by: Christian Ang <angc@vmware.com>
